### PR TITLE
Normalize archives before instrumenting

### LINKED ISSUE #14897: Gradle should reuse files in jars-8 instead of producing duplicates
### Expected Behavior

Gradle should reuse the files in `.gradle/caches` instead of producing duplicates.

### Current Behavior

```
$ cd $HOME/.gradle/caches/jars-8
$ find . -name '*.jar' -print0 | xargs -0 shasum | sort
```

```
bcd346ac4a39f859d7cbee2b7a80ada3b4e0957a  ./0473060bc3b78553671ddb50e49f3a79/fmpp.jar
bcd346ac4a39f859d7cbee2b7a80ada3b4e0957a  ./23828adeb529761effa32c2fb5a8a9f7/fmpp.jar
bcd346ac4a39f859d7cbee2b7a80ada3b4e0957a  ./2677af150555032a72a722411fbc5a3b/fmpp.jar
bcd346ac4a39f859d7cbee2b7a80ada3b4e0957a  ./5e6f7ddce3d43736875e403d769cd87f/fmpp.jar
bcd346ac4a39f859d7cbee2b7a80ada3b4e0957a  ./9979c5fb5ce3e7e041193dcaa8464bb5/fmpp.jar
bcd346ac4a39f859d7cbee2b7a80ada3b4e0957a  ./a8083af00eb3b857b0618f7f1339363c/fmpp.jar
bcd346ac4a39f859d7cbee2b7a80ada3b4e0957a  ./be419036a54cc47c8aca3debee19ccff/fmpp.jar
bcd346ac4a39f859d7cbee2b7a80ada3b4e0957a  ./c99ee5c728ce1a670ff73e3909d0a31e/fmpp.jar
bcd346ac4a39f859d7cbee2b7a80ada3b4e0957a  ./ceccdeb8791d4ae1083d391c4f77fc82/fmpp.jar
bcd346ac4a39f859d7cbee2b7a80ada3b4e0957a  ./eade2c30c2ad5de139bb025b22355dbf/fmpp.jar
bcd346ac4a39f859d7cbee2b7a80ada3b4e0957a  ./f04dd9f28eceb8d02c89da5111005440/fmpp.jar
....
f9a6da8939fcfa673c59373f8a13a49b8978300d  ./275b6159d7a94e8ad48fb91828419fb3/buildSrc.jar
f9a6da8939fcfa673c59373f8a13a49b8978300d  ./47781faa30176e966a63c2878da3263f/buildSrc.jar
f9a6da8939fcfa673c59373f8a13a49b8978300d  ./4afc89916856edf38e7b2b9da87cfca7/buildSrc.jar
f9a6da8939fcfa673c59373f8a13a49b8978300d  ./7071d2c65ca29a82d43dd577c1741120/buildSrc.jar
f9a6da8939fcfa673c59373f8a13a49b8978300d  ./75bcfa9e1d47833ecd4defb782b5120c/buildSrc.jar
f9a6da8939fcfa673c59373f8a13a49b8978300d  ./977bc72efcf49411832791fb573c44d2/buildSrc.jar
f9a6da8939fcfa673c59373f8a13a49b8978300d  ./9b7487e3eb891dea094cdb590dc5e6f7/buildSrc.jar
f9a6da8939fcfa673c59373f8a13a49b8978300d  ./af357574b75096fde52f05bd5ca15c19/buildSrc.jar
f9a6da8939fcfa673c59373f8a13a49b8978300d  ./c00150b6df94613fc7594ede1bc1a7ad/buildSrc.jar
f9a6da8939fcfa673c59373f8a13a49b8978300d  ./de26e4bbf87dea79096b2851d5f39ee3/buildSrc.jar
f9a6da8939fcfa673c59373f8a13a49b8978300d  ./ea76618b9626aa8049a4a0c21c3df4ae/buildSrc.jar
f9a6da8939fcfa673c59373f8a13a49b8978300d  ./f2a176b0e7b65f9dc0a9df71c67e8049/buildSrc.jar
...
ffb99681934ef6b54cc954612322d87eb0a3883f  ./5d0789cc7b5a061666377c7f68481d9c/gson-2.8.6.jar
ffb99681934ef6b54cc954612322d87eb0a3883f  ./78c2d0c1036b4b9ee697381a6efd49a0/gson-2.8.6.jar
```

### Context

I'm trying to cache `.gradle` folder at Travis, and 

### Steps to Reproduce 

Execute `gradlew rat` task in https://github.com/apache/calcite
For instance, `fmpp.jar` comes from https://github.com/apache/calcite/tree/cd922deff8ae3f25546b1d77fb147c3098eb177b/buildSrc/subprojects/fmpp

### Your Environment

Gradle 6.7

### PR DESCRIPTION
**Pull request by mlopatkin**
Fixes #14897

### CODE REVIEW

**Comment by mlopatkin** [comment]
@bot-gradle test APT

**Comment by mlopatkin** [comment]
@bot-gradle help

**Comment by mlopatkin** [comment]
@bot-gradle test PT

**Comment by mlopatkin** [comment]
@bot-gradle test PT

**Comment by mlopatkin** [comment]
@bot-gradle test this

File: subprojects/core/src/main/java/org/gradle/internal/classpath/InstrumentingClasspathFileTransformer.java [hunk 1]
```diff
@@ -112,19 +112,28 @@ public InstrumentingClasspathFileTransformer(
...
         CachedClasspathTransformer.Transform transform,
         GradleCoreInstrumentingTypeRegistry gradleCoreInstrumentingTypeRegistry
     ) {
         this.fileLockManager = fileLockManager;
         this.classpathWalker = classpathWalker;
         this.classpathBuilder = classpathBuilder;
+
+        HashCode configHash = configHashFor(policy, transform, gradleCoreInstrumentingTypeRegistry);
+        this.fileHasher = sourceSnapshot -> {
+            Hasher hasher = Hashing.defaultFunction().newHasher();
```
Comment (bamboo) [inline subprojects/core/src/main/java/org/gradle/internal/classpath/InstrumentingClasspathFileTransformer.java:126]: 💅 Extract this block to a method to keep the constructor simple and mainly about assigning fields

File: subprojects/core/src/test/groovy/org/gradle/internal/classpath/InstrumentingClasspathFileTransformerTest.groovy [hunk 2]
```diff
@@ -54,6 +55,7 @@ class InstrumentingClasspathFileTransformerTest extends Specification {
     }
     def typeRegistry = Stub(InstrumentingTypeRegistry)
 
+
```
Comment (bamboo) [inline subprojects/core/src/test/groovy/org/gradle/internal/classpath/InstrumentingClasspathFileTransformerTest.groovy:58]: ❓ Is the additional empty line here intentional?

File: subprojects/core/src/integTest/groovy/org/gradle/internal/classpath/BuildScriptClasspathIntegrationSpec.groovy [hunk 3]
```diff
@@ -477,5 +521,18 @@ class BuildScriptClasspathIntegrationSpec extends AbstractIntegrationSpec implem
...
 
-
+    /**
+     * Finds all cached transformed JARs named {@code jarName}.
+     * @param jarName the name of the JAR to look
+     * @return the list of transformed JARs in the cache
+     */
+    List<File> getCachedTransformedJarsByName(String jarName) {
+        Arrays.stream(cacheDir.listFiles()).filter {
+            File cacheChild -> !cacheChild.name.startsWith("o_") && cacheChild.isDirectory()
```
Comment (bamboo) [inline subprojects/core/src/integTest/groovy/org/gradle/internal/classpath/BuildScriptClasspathIntegrationSpec.groovy:531]: 💅 A comment about the relevance of the `"o_"` prefix would be useful here

File: subprojects/internal-integ-testing/src/main/groovy/org/gradle/integtests/fixtures/executer/GradleBackedArtifactBuilder.java [hunk 4]
```diff
@@ -51,11 +53,19 @@ public void manifestAttributes(Map<String, String> attributes) {
...
+        }
+        shouldPreserveTimestamps = preserveTimestamps;
+    }
+
     @Override
     public void buildJar(File jarFile) {
-        String conf = executer.getDistribution().getVersion().compareTo(GradleVersion.version("3.4")) < 0 ? "compile" : "implementation";
-        String destinationDir = executer.getDistribution().getVersion().compareTo(GradleVersion.version("5.1")) < 0 ? "destinationDir" : "destinationDirectory";
-        String archiveName = executer.getDistribution().getVersion().compareTo(GradleVersion.version("5.0")) < 0 ? "archiveName" : "archiveFileName";
+        String conf = isGradleExecuterVersionLessThan("3.4") ? "compile" : "implementation";
```
Comment (bamboo) [inline subprojects/internal-integ-testing/src/main/groovy/org/gradle/integtests/fixtures/executer/GradleBackedArtifactBuilder.java:66]: 👍

File: subprojects/core/src/integTest/groovy/org/gradle/internal/classpath/BuildScriptClasspathIntegrationSpec.groovy [hunk 5]
```diff
@@ -459,6 +461,48 @@ class BuildScriptClasspathIntegrationSpec extends AbstractIntegrationSpec implem
...
+        '''
+        def reproducibleJar = file("reproducible/testClasses.jar")
+        def currentTimestampJar = file("current/testClasses.jar")
+        artifactBuilder().tap {
+            preserveTimestamps(false)
+            sourceFile("org/gradle/test/BuildClass.java").text = buildClassSource
+            buildJar(reproducibleJar)
+        }
+        artifactBuilder().tap {
+            preserveTimestamps(true)
```
Comment (bamboo) [inline subprojects/core/src/integTest/groovy/org/gradle/internal/classpath/BuildScriptClasspathIntegrationSpec.groovy:482]: 👍

**Review by bamboo — APPROVED** [review]

**Pull request merged**